hicexplorer/hicBuildMatrix.py

```diff
@@ -822,74 +844,363 @@ def main(args=None):
+    str1 = pysam.Samfile(args.samFiles[0].name, 'rb')
+    str2 = pysam.Samfile(args.samFiles[1].name, 'rb')
+
+    args.samFiles[0].close()
+    args.samFiles[1].close()
+    outputFileBufferDir = args.outputFileBufferDir
+    if not outputFileBufferDir[:-1] == '/':
+        outputFileBufferDir += '/'
+    if args.outBam:
+        args.outBam.close()
+
+    chrom_sizes = get_chrom_sizes(str1)
+
+    read_pos_matrix = ReadPositionMatrix()
+
+    # define bins
+    rf_positions = None
+    if args.restrictionCutFile:
+        rf_interval = bed2interval_list(args.restrictionCutFile)
+        bin_intervals = get_rf_bins(rf_interval,
+                                    min_distance=args.minDistance,
+                                    max_distance=args.maxDistance)
+
+        rf_positions = intervalListToIntervalTree(rf_interval)
     else:
-        hic_matrix += coo_matrix((data, (row, col)), shape=(matrix_size, matrix_size))
+        bin_intervals = get_bins(args.binSize, chrom_sizes, args.region)
+
+    matrix_size = len(bin_intervals)
+    bin_intval_tree = intervalListToIntervalTree(bin_intervals)
+    ref_id2name = str1.references
+
+    # build c_type shared memory for the interval tree
+    shared_array_list = []
+    index_dict = {}
+    j = 0
+    interval_list = []
+    for i, seq in enumerate(bin_intval_tree):
+        start = j
+        interval_list = []
+        for interval in bin_intval_tree[seq]:
+            interval_list.append((interval.begin, interval.end, interval.data))
+            j += 1
+        end = j - 1
+        index_dict[seq] = (start, end)
+        interval_list = sorted(interval_list)
+        shared_array_list.extend(interval_list)
+    shared_build_intval_tree = RawArray(C_Interval, shared_array_list)
+    bin_intval_tree = None
+    dangling_sequences = dict()
+    if args.restrictionSequence:
+        # build a list of dangling sequences
+        args.restrictionSequence = args.restrictionSequence.upper()
+        dangling_sequences['pat_forw'] = args.restrictionSequence[1:]
+        dangling_sequences['pat_rev'] = args.restrictionSequence[:-1]
+        sys.stderr.write("dangling sequences to check "
+                         "are {}\n".format(dangling_sequences))
+
+    # initialize coverage vectors that
+    # save the number of reads that overlap
+    # a bin.
+    # To save memory, coverage is not measured by bp
+    # but by bins of length 10bp
+    binsize = 10
+    number_of_elements_coverage = 0
+    start_pos_coverage = []
+    end_pos_coverage = []
+
+    for value in bin_intervals:
+        chrom, start, end = value
+        start_pos_coverage.append(number_of_elements_coverage)
+
+        number_of_elements_coverage += (end - start) / binsize
+        end_pos_coverage.append(number_of_elements_coverage - 1)
+    pos_coverage_ = zip(start_pos_coverage, end_pos_coverage)
+    pos_coverage = RawArray(C_Coverage, pos_coverage_)
+    pos_coverage_ = None
```

Comment:
To have the start and end positions for each chromosome in the 1D-Array?
Or do you mean that it could be written as: ```pos_coverage = RawArray(C_Coverage, zip(start_pos_coverage, end_pos_coverage))```